k2/csrc/fsa_algo.cu

```diff
@@ -288,4 +289,127 @@ void ArcSort(Fsa &src, Fsa *dest, Array1<int32_t> *arc_map /*= nullptr*/) {
+
+  int32_t num_arcs = static_cast<int32_t>(tmp_arc_indexes.size());
+
+  const Arc *src_arcs_data = src.values.Data();
+  Array1<Arc> arcs(context, num_arcs);
+  Arc *arcs_data = arcs.Data();
+  int32_t cur_state = 0;
+  for (auto i : tmp_arc_indexes) {
+    const Arc &src_arc = src_arcs_data[i];
+    arcs_data[cur_state] =
+        Arc(cur_state, cur_state + 1, src_arc.label, src_arc.score);
+    cur_state += 1;
+  }
+  std::vector<int32_t> row_splits_vec(cur_state + 2);
+  std::iota(row_splits_vec.begin(), row_splits_vec.end(), 0);
+  row_splits_vec.back() -= 1;
+  Array1<int32_t> row_splits(context, row_splits_vec);
+  RaggedShape shape = RaggedShape2(&row_splits, nullptr, num_arcs);
+  *out = Fsa(shape, arcs);
+
+  if (best_path_arc_indexes != nullptr)
+    *best_path_arc_indexes = Array1<int32_t>(context, tmp_arc_indexes);
+
+  return state_weights.Back();
+}
+
+Ragged<int32_t> ShortestPath(FsaVec &fsas,
+                             Array1<float> *total_scores /*= nullptr*/,
+                             Array1<int32_t> *entering_arcs /*= nullptr*/) {
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+  Ragged<int32_t> state_batches = GetStateBatches(fsas, true);
+  Array1<int32_t> dest_states = GetDestStates(fsas, true);
+  Ragged<int32_t> incoming_arcs = GetIncomingArcs(fsas, dest_states);
+  Ragged<int32_t> entering_arc_batches =
+      GetEnteringArcIndexBatches(fsas, incoming_arcs, state_batches);
+
+  bool log_semiring = false;
+  Array1<int32_t> tmp_entering_arcs;
+  Array1<float> forward_scores =
+      GetForwardScores<float>(fsas, state_batches, entering_arc_batches,
+                              log_semiring, &tmp_entering_arcs);
+  if (total_scores != nullptr)
+    *total_scores = GetTotScores<float>(fsas, forward_scores);
+
+  if (entering_arcs != nullptr) *entering_arcs = tmp_entering_arcs;
+
+  const int32_t *entering_arcs_data = tmp_entering_arcs.Data();
+  const Arc *arcs_data = fsas.values.Data();
+  int32_t num_fsas = fsas.Dim0();
+  ContextPtr &context = fsas.Context();
+
+  // allocate an extra element for ExclusiveSum
+  Array1<int32_t> num_best_arcs_per_fsa(context, num_fsas + 1);
+  int32_t *num_best_arcs_per_fsa_data = num_best_arcs_per_fsa.Data();
+  const int32_t *row_splits1_data = fsas.RowSplits(1).Data();
+
+  auto lambda_set_num_best_arcs = [=] __host__ __device__(int32_t fsas_idx0) {
+    int32_t state_idx01 = row_splits1_data[fsas_idx0];
+    int32_t state_idx01_next = row_splits1_data[fsas_idx0 + 1];
+
+    if (state_idx01_next == state_idx01) {
+      // this fsa is empty, so there is no best path available
+      num_best_arcs_per_fsa_data[fsas_idx0] = 0;
+      return;
+    }
+
+    int32_t final_state_idx01 = state_idx01_next - 1;
+    int32_t cur_state = final_state_idx01;
+    int32_t cur_index = entering_arcs_data[cur_state];
+    int32_t num_arcs = 0;
+    while (cur_index != -1) {
+      cur_state = arcs_data[cur_index].src_state + state_idx01;
```

Comment:
I don't like that you have 2 kernels that both have a loop.  You could, in this kernel, store the best traceback in a ragged array that was "too large" (e.g. had the total num-states in the input FSAs), and then copy the info to the output within a kernel that was loop-free.

Reply:
I'm OK to merge this anyway, but please put it on your TODO list to fix this.

Reply:
Will change it as you are willing to trade space for speed.

Reply:
Fixed.

Reply:
It still has a loop that iterates over the number of states.